File: C++/Complex Windows/Complex Testing Console/Complex Testing Console.cpp
﻿/*
 *   Copyright 2022 Edmond Chow
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   disTributed under the License is disTributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */
#include "Base.h"
#include "Complex.h"
#define Gbl __stdcall
using namespace ComplexTestingConsole;
using namespace Num;
namespace CmplxBasis
{
	template <typename T>
	void Mul(const std::wstring& L, const wchar_t* R, T(Gbl* S)(const Cmplx&, const Cmplx&))
	{
		if (L == R)
		{
			Cmplx U = Cmplx::Val(Base::Input(L"U = "));
			Cmplx V = Cmplx::Val(Base::Input(L"V = "));
			T Rst = std::invoke(S, U, V);
			std::wstring Str;
			if constexpr (std::is_same_v<T, Vec1D>) { Str = ToModStr(Cmplx{ 0, Rst }); }
			else { Str = ToModStr(Rst); }
			Base::Output(Str);
		}
	};
	template <typename T>
	void Op(const std::wstring& L, const wchar_t* R, T(Gbl* S)(const Cmplx&, const Cmplx&))
	{
		if (L == R)
		{
			Cmplx U = Cmplx::Val(Base::Input(L"U = "));
			Cmplx V = Cmplx::Val(Base::Input(L"V = "));
			Base::Output(ToModStr(std::invoke(S, U, V)));
		}
	};
	inline void Pow(const std::wstring& L, const wchar_t* R, Cmplx(Gbl* S)(const Cmplx&, std::int64_t))
	{
		if (L == R)
		{
			Cmplx U = Cmplx::Val(Base::Input(L"U = "));
			std::int64_t V = AsInt(Base::Input(L"V = "));
			Base::Output(ToModStr(std::invoke(S, U, V)));
		}
	};
	template <typename... As>
	void Pow(const std::wstring& L, std::wstring&& R, Cmplx(Gbl* S)(const Cmplx&, const Cmplx&, std::int64_t, As...))
	{
		if (L == R)
		{
			Cmplx U = Cmplx::Val(Base::Input(L"U = "));
			Cmplx V = Cmplx::Val(Base::Input(L"V = "));
			std::array<std::int64_t, 1 + sizeof...(As)> Dat{};
			PowGet(Dat);
			PowRst(S, U, V, Dat);
		}
		else if (L == R + L"()")
		{
			Cmplx U = Cmplx::Val(Base::Input(L"U = "));
			Cmplx V = Cmplx::Val(Base::Input(L"V = "));
			std::array<std::pair<std::int64_t, std::int64_t>, 1 + sizeof...(As)> Dat{};
			PowGet(Dat);
			PowRst(S, R, U, V, Dat);
		}
	};
	template <typename T>
	void Bas(const std::wstring& L, const wchar_t* R, T(Gbl* S)(const Cmplx&))
	{
		if (L == R)
		{
			Cmplx V = Cmplx::Val(Base::Input(L"V = "));
			Base::Output(ToModStr(std::invoke(S, V)));
		}
	};
	template <typename T>
	void Bas(const std::wstring& L, std::wstring&& R, T(Gbl* S)(const Cmplx&, std::int64_t))
	{
		if (L == R)
		{
			Cmplx V = Cmplx::Val(Base::Input(L"V = "));
			std::int64_t P = AsInt(Base::Input(L"P = "));
			Base::Output(ToModStr(std::invoke(S, V, P)));
		}
		else if (L == R + L"()")
		{
			Cmplx V = Cmplx::Val(Base::Input(L"V = "));
			std::int64_t PMin = AsInt(Base::Input(L"P(min) = "));
			std::int64_t PMax = AsInt(Base::Input(L"P(max) = "));
			for (std::int64_t P = PMin; P <= PMax; ++P)
			{
				Base::Output(R + L"(" + ToModStr(P) + L") = ", ToModStr(std::invoke(S, V, P)));
			}
		}
	};
	inline void Tri(const std::wstring& L, const wchar_t* R, Cmplx(Gbl* S)(const Cmplx&))
	{
		if (L == R)
		{
			Cmplx V = Cmplx::Val(Base::Input(L"V = "));
			Base::Output(ToModStr(std::invoke(S, V)));
		}
	};
	inline void Atri(const std::wstring& L, std::wstring&& R, Cmplx(Gbl* S)(const Cmplx&, bool, std::int64_t))
	{
		if (L == R)
		{
			Cmplx V = Cmplx::Val(Base::Input(L"V = "));
			bool Sign = false;
			std::wstring Input = std::regex_replace(Base::Input(L"Sign : "), std::wregex(L" "), L"");
			if (Input == L"+") { Sign = true; }
			else if (Input != L"-") { throw std::invalid_argument("A String interpretation of the sign cannot be converted as a bool value."); }
			std::int64_t P = AsInt(Base::Input(L"P = "));
			Base::Output(ToModStr(std::invoke(S, V, Sign, P)));
		}
		else if (L == R + L"()")
		{
			Cmplx V = Cmplx::Val(Base::Input(L"V = "));
			std::int64_t PMin = AsInt(Base::Input(L"P(min) = "));
			std::int64_t PMax = AsInt(Base::Input(L"P(max) = "));
			for (std::int64_t P = PMin; P <= PMax; ++P)
			{
				Base::Output(R + L"(+, " + ToModStr(P) + L") = ", ToModStr(std::invoke(S, V, true, P)));
			}
			for (std::int64_t P = PMin; P <= PMax; ++P)
			{
				Base::Output(R + L"(-, " + ToModStr(P) + L") = ", ToModStr(std::invoke(S, V, false, P)));
			}
		}
	};
	class CmplxConsole final
	{
	private:
		constexpr CmplxConsole() noexcept = delete;
		constexpr ~CmplxConsole() noexcept = delete;
	public:
		static void Load() noexcept;
	};
	void CmplxConsole::Load() noexcept
	{
		Base::Startup(Base::GetTitle());
		Base::Selection(L"=   +   -   *   /   ^   power()   root()   log()");
		Base::Selection(L"abs   arg()   conjg   sgn   inverse   exp   ln()   dot   outer   even   cross");
		Base::Selection(L"sin   cos   tan   csc   sec   cot   arcsin()   arccos()   arctan()   arccsc()   arcsec()   arccot()");
		Base::Selection(L"sinh   cosh   tanh   csch   sech   coth   arcsinh()   arccosh()   arctanh()   arccsch()   arcsech()   arccoth()");
		Base::Selection(Base::GetStartupLine());
		for (std::wstring L; !Base::IsSwitchTo(L); L = Base::Input())
		{
			if (L.empty()) { continue; }
			try
			{
				Op(L, L"=", operator ==);
				Op(L, L"+", operator +);
				Op(L, L"-", operator -);
				Op(L, L"*", operator *);
				Op(L, L"/", operator /);
				/****/
				Pow(L, L"^", operator ^);
				Pow(L, L"power", Cmplx::Power);
				Pow(L, L"root", Cmplx::Root);
				Pow(L, L"log", Cmplx::Log);
				/****/
				Mul(L, L"dot", Cmplx::Dot);
				Mul(L, L"outer", Cmplx::Outer);
				Mul(L, L"even", Cmplx::Even);
				Mul(L, L"cross", Cmplx::Cross);
				/****/
				Bas(L, L"abs", Cmplx::Abs);
				Bas(L, L"arg", Cmplx::Arg);
				Bas(L, L"conjg", Cmplx::Conjg);
				Bas(L, L"sgn", Cmplx::Sgn);
				Bas(L, L"inverse", Cmplx::Inverse);
				Bas(L, L"exp", Cmplx::Exp);
				Bas(L, L"ln", Cmplx::Ln);
				/****/
				Tri(L, L"sin", Cmplx::Sin);
				Tri(L, L"cos", Cmplx::Cos);
				Tri(L, L"tan", Cmplx::Tan);
				Tri(L, L"csc", Cmplx::Csc);
				Tri(L, L"sec", Cmplx::Sec);
				Tri(L, L"cot", Cmplx::Cot);
				Tri(L, L"sinh", Cmplx::Sinh);
				Tri(L, L"cosh", Cmplx::Cosh);
				Tri(L, L"tanh", Cmplx::Tanh);
				Tri(L, L"csch", Cmplx::Csch);
				Tri(L, L"sech", Cmplx::Sech);
				Tri(L, L"coth", Cmplx::Coth);
				Atri(L, L"arcsin", Cmplx::Arcsin);
				Atri(L, L"arccos", Cmplx::Arccos);
				Atri(L, L"arctan", Cmplx::Arctan);
				Atri(L, L"arccsc", Cmplx::Arccsc);
				Atri(L, L"arcsec", Cmplx::Arcsec);
				Atri(L, L"arccot", Cmplx::Arccot);
				Atri(L, L"arcsinh", Cmplx::Arcsinh);
				Atri(L, L"arccosh", Cmplx::Arccosh);
				Atri(L, L"arctanh", Cmplx::Arctanh);
				Atri(L, L"arccsch", Cmplx::Arccsch);
				Atri(L, L"arcsech", Cmplx::Arcsech);
				Atri(L, L"arccoth", Cmplx::Arccoth);
			}
			catch (const std::exception& Ex) { Base::Exception(Ex); }
		}
	};
}

File: C#/Complex Windows/Complex Testing Console/Complex Testing Console.cs
/*
 *   Copyright 2022 Edmond Chow
 *   
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *   
 *       http://www.apache.org/licenses/LICENSE-2.0
 *   
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */
using System;
using System.Reflection;
using ComplexTestingConsole;
using static ComplexTestingConsole.Module;
using Num;
using static Num.Cmplx;
using static CmplxBasis;
internal static class CmplxBasis
{
    internal static void Mul<T>(string L, string R, Func<Cmplx, Cmplx, T> F)
    {
        if (L == R)
        {
            Cmplx U = Val(Base.Input("U = "));
            Cmplx V = Val(Base.Input("V = "));
            object Rst = F.Invoke(U, V);
            if (Rst is Vec1D Ve) { Rst = new Cmplx(0, Ve); }
            Base.Output(Rst.ToModStr());
        }
    }
    internal static void Op<T>(string L, string R, Func<Cmplx, Cmplx, T> F)
    {
        if (L == R)
        {
            Cmplx U = Val(Base.Input("U = "));
            Cmplx V = Val(Base.Input("V = "));
            Base.Output(F.Invoke(U, V).ToModStr());
        }
    }
    internal static void PowOp(string L, string R, Func<Cmplx, long, Cmplx> F)
    {
        if (L == R)
        {
            Cmplx U = Val(Base.Input("U = "));
            long V = Base.Input("V = ").Int();
            Base.Output(F.Invoke(U, V).ToModStr());
        }
    }
    internal static void Pow(string L, string R, Func<Cmplx, Cmplx, long, Cmplx> F)
    {
        Pow(L, R, Delegate.CreateDelegate(F.GetType(), F.GetMethodInfo()));
    }
    internal static void Pow(string L, string R, Func<Cmplx, Cmplx, long, long, Cmplx> F)
    {
        Pow(L, R, Delegate.CreateDelegate(F.GetType(), F.GetMethodInfo()));
    }
    internal static void Pow(string L, string R, Delegate F)
    {
        if (L == R)
        {
            Cmplx U = Val(Base.Input("U = "));
            Cmplx V = Val(Base.Input("V = "));
            object[] Args = new object[F.GetMethodInfo().GetParameters().Length];
            Args[0] = U;
            Args[1] = V;
            PowGet(Args);
            PowRst(F, Args);
        }
        else if (L == R + "()")
        {
            Cmplx U = Val(Base.Input("U = "));
            Cmplx V = Val(Base.Input("V = "));
            object[] Args = new object[F.GetMethodInfo().GetParameters().Length];
            long[] Upper = new long[Args.Length - 2];
            Args[0] = U;
            Args[1] = V;
            PowGet(Args, Upper);
            PowRst(F, R, Args, Upper);
        }
    }
    internal static void Bas<T>(string L, string R, Func<Cmplx, T> F)
    {
        if (L == R)
        {
            Cmplx V = Val(Base.Input("V = "));
            Base.Output(F.Invoke(V).ToModStr());
        }
    }
    internal static void BasP<T>(string L, string R, Func<Cmplx, long, T> F)
    {
        if (L == R)
        {
            Cmplx V = Val(Base.Input("V = "));
            long P = Base.Input("P = ").Int();
            Base.Output(F.Invoke(V, P).ToModStr());
        }
        else if (L == R + "()")
        {
            Cmplx V = Val(Base.Input("V = "));
            long PMin = Base.Input("P(min) = ").Int();
            long PMax = Base.Input("P(max) = ").Int();
            for (long P = PMin; P <= PMax; ++P)
            {
                Base.Output(R + "(" + P.ToModStr() + ") = ", F.Invoke(V, P).ToModStr());
            }
        }
    }
    internal static void Tri(string L, string R, Func<Cmplx, Cmplx> F)
    {
        if (L == R)
        {
            Cmplx V = Val(Base.Input("V = "));
            Base.Output(F.Invoke(V).ToModStr());
        }
    }
    internal static void Atri(string L, string R, Func<Cmplx, bool, long, Cmplx> F)
    {
        if (L == R)
        {
            Cmplx V = Val(Base.Input("V = "));
            bool S = false;
            string Ipt = Base.Input("Sign : ").Replace(" ", "");
            if (Ipt == "+") { S = true; }
            else if (Ipt != "-") { throw new ArgumentException("A string interpretation of the sign cannot be converted as a bool value."); }
            long P = Base.Input("P = ").Int();
            Base.Output(F.Invoke(V, S, P).ToModStr());
        }
        else if (L == R + "()")
        {
            Cmplx V = Val(Base.Input("V = "));
            long PMin = Base.Input("P(min) = ").Int();
            long PMax = Base.Input("P(max) = ").Int();
            for (long P = PMin; P <= PMax; ++P)
            {
                Base.Output(R + "(+, " + P.ToModStr() + ") = ", F.Invoke(V, true, P).ToModStr());
            }
            for (long P = PMin; P <= PMax; ++P)
            {
                Base.Output(R + "(-, " + P.ToModStr() + ") = ", F.Invoke(V, false, P).ToModStr());
            }
        }
    }
}
internal static class CmplxConsole
{
    internal static void Load()
    {
        Base.Startup(Base.GetTitle());
        Base.Selection("=   +   -   *   /   ^   power()   root()   log()");
        Base.Selection("abs   arg()   conjg   sgn   inverse   exp   ln()   dot   outer   even   cross");
        Base.Selection("sin   cos   tan   csc   sec   cot   arcsin()   arccos()   arctan()   arccsc()   arcsec()   arccot()");
        Base.Selection("sinh   cosh   tanh   csch   sech   coth   arcsinh()   arccosh()   arctanh()   arccsch()   arcsech()   arccoth()");
        Base.Selection(Base.GetStartupLine());
        for (string L = ""; !Base.IsSwitchTo(L); L = Base.Input())
        {
            if (L == "") { continue; }
            try
            {
                Op(L, "=", (Cmplx U, Cmplx V) => { return U == V; });
                Op(L, "+", (Cmplx U, Cmplx V) => { return U + V; });
                Op(L, "-", (Cmplx U, Cmplx V) => { return U - V; });
                Op(L, "*", (Cmplx U, Cmplx V) => { return U * V; });
                Op(L, "/", (Cmplx U, Cmplx V) => { return U / V; });
                /****/
                PowOp(L, "^", (Cmplx U, long V) => { return U ^ V; });
                Pow(L, "power", Power);
                Pow(L, "root", Root);
                Pow(L, "log", Log);
                /****/
                Mul(L, "dot", Dot);
                Mul(L, "outer", Outer);
                Mul(L, "even", Even);
                Mul(L, "cross", Cross);
                /****/
                Bas(L, "abs", Abs);
                BasP(L, "arg", Arg);
                Bas(L, "conjg", Conjg);
                Bas(L, "sgn", Sgn);
                Bas(L, "inverse", Inverse);
                Bas(L, "exp", Exp);
                BasP(L, "ln", Ln);
                /****/
                Tri(L, "sin", Sin);
                Tri(L, "cos", Cos);
                Tri(L, "tan", Tan);
                Tri(L, "csc", Csc);
                Tri(L, "sec", Sec);
                Tri(L, "cot", Cot);
                Tri(L, "sinh", Sinh);
                Tri(L, "cosh", Cosh);
                Tri(L, "tanh", Tanh);
                Tri(L, "csch", Csch);
                Tri(L, "sech", Sech);
                Tri(L, "coth", Coth);
                Atri(L, "arcsin", Arcsin);
                Atri(L, "arccos", Arccos);
                Atri(L, "arctan", Arctan);
                Atri(L, "arccsc", Arccsc);
                Atri(L, "arcsec", Arcsec);
                Atri(L, "arccot", Arccot);
                Atri(L, "arcsinh", Arcsinh);
                Atri(L, "arccosh", Arccosh);
                Atri(L, "arctanh", Arctanh);
                Atri(L, "arccsch", Arccsch);
                Atri(L, "arcsech", Arcsech);
                Atri(L, "arccoth", Arccoth);
            }
            catch (Exception Ex) { Base.Exception(Ex); }
        }
    }
}
